src/io/proxi.rs

```diff
@@ -1,12 +1,231 @@
-use std::fmt::Display;
-use std::str::FromStr;
+use std::{cmp::Ordering, fmt::Display, marker::PhantomData, str::FromStr};
 
+use num_traits::AsPrimitive;
 use serde::{Deserialize, Serialize};
 
-use super::usi::USI;
-use crate::params::{ControlledVocabulary, Param, ParamCow, Value, CURIE};
-use crate::spectrum::{ArrayType, BinaryArrayMap, BinaryDataArrayType, DataArray, IsolationWindowState, MultiLayerSpectrum, Precursor, ScanPolarity, SignalContinuity, SpectrumDescription};
-use crate::{curie, prelude::*};
+use crate::{
+    curie,
+    io::usi::USI,
+    params::{ControlledVocabulary, Param, ParamCow, Value, CURIE},
+    prelude::*,
+    spectrum::{
+        ArrayType, BinaryArrayMap, BinaryDataArrayType, DataArray, IsolationWindowState,
+        MultiLayerSpectrum, Precursor, ScanPolarity, SignalContinuity, SpectrumDescription,
+    },
+};
+
+/// The possible PROXI backends
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+pub enum PROXIBackend {
+    PeptideAtlas,
+    MassIVE,
+    Pride,
+    Jpost,
+    ProteomeXchange,
+}
+
+impl PROXIBackend {
+    const ALL: &[Self] = &[
+        Self::PeptideAtlas,
+        Self::MassIVE,
+        Self::Pride,
+        Self::Jpost,
+        Self::ProteomeXchange,
+    ];
+
+    /// The PROXI server base url which needs concatenating of the USI at the end
+    const fn base_url(self) -> &'static str {
+        match self {
+            Self::PeptideAtlas => "http://www.peptideatlas.org/api/proxi/v0.1/spectra?resultType=full&usi=",
+            Self::MassIVE => "http://massive.ucsd.edu/ProteoSAFe/proxi/v0.1/spectra?resultType=full&usi=",
+            Self::Pride => "http://www.ebi.ac.uk/pride/proxi/archive/v0.1/spectra?resultType=full&usi=",
+            Self::Jpost => "https://repository.jpostdb.org/proxi/spectra?resultType=full&usi=",
+            Self::ProteomeXchange => "http://proteomecentral.proteomexchange.org/api/proxi/v0.1/spectra?resultType=full&usi=",
+        }
+    }
+}
+
+impl USI {
+    /// Retrieve this USI from the given PROXI backend. If no PROXI backend is indicated it will
+    /// aggregate the results from all known backends and return the first successful spectrum.
+    ///
+    /// This function is only available with the feature `proxi`.
+    pub fn get_spectrum_blocking(
```

Comment:
Stylistically I'm tempted to ask this be renamed to something more succinct but "`proxi`" is meaningless to many new users and I've no better ideas.

Reply:
I fully agree, but also did not find a better name yet. Maybe 'retrieve'/'fetch'/'load' with or without the 'spectrum' part?

Reply:
Since I wrote the `pyteomics` implementation, I was partial to that here too. PROXI was still being worked on in recent memory way back then too. The mzSpecLib project has taken all the attention away from PROXI since then, and it's on the "todo" list to switch the PROXI JSON format to mirror the mzSpecLib JSON format.

Reply:
Reading a bit more into the PROXI standard I do agree that calling it .proxi is not the best name, if only for the fact that proxi allows many more types of calls to the server. Keeping it on `get_spectrum` seems fine to me, that is also roughly the name of the proxi 'function' that is called in the function. Although the 'get' part could be swapped out to indicate the network request a bit more, something like 'download_spectrum'  might be more clear to anyone reading that it does network requests.